Try: React Components in the Admin Script

jblz opened the pull request:
WIP!

This is merely a demo of how we can use react components in the admin page.

jblz committed:
Try: React Components in the Admin Script

jblz committed:
Try: REST API Endpoint for Settings
Make get_options public

jase-d-ace committed:
Add static folder for services file

jase-d-ace committed:
Add components folder

jase-d-ace committed:
Render react component instead of static header

jase-d-ace committed:
create setting component

jase-d-ace committed:
import new component and start thinking about render logic

jase-d-ace committed:
transform json into array before passing to component

jase-d-ace committed:
fix render logic for settings

jase-d-ace committed:
create dynamic input component

jase-d-ace committed:
pass props to new component

jase-d-ace committed:
create select and option components

jase-d-ace committed:
import and use select and option components

jase-d-ace committed:
change variable name

jase-d-ace committed:
make checkboxes dynamic

jase-d-ace committed:
pass json instead of array to components

jase-d-ace committed:
render all settings and add form interactivity

jase-d-ace committed:
add submit button to react form

jase-d-ace committed:
remove console log and also start thinking about form submit

jase-d-ace commented:
I hit MVP of a VERY barebones, CSSless React app that emulates the current form's fields and updates state to track user input (but the current form still appears underneath, which is kinda funny to see). 

Is there anything else I need on there before I click the "Ready for Review" button?

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into remove/jquery-dependency--try-react

jblz committed:
update build

jblz committed:
install @wordpress/element

jblz committed:
lint fixes & formatting

jblz commented:
@jase-d-ace -- I rebased this against the latest work in `develop`.  I hope I didn't mangle it too badly!

jblz committed:
Add export script (#316)

* Script to export plugin to zip or dir

* Add a comment to the .gitattributes file

* export script: assume remote is origin & wrap for fetch

* Tidy up branch name to rm illegal chars for file names

* allow dots in fname

GaryJones committed:
CI: Use Ubuntu 18.04 (#315)

Ubuntu 16.04 is deprecated, but setup-php has been updated to make using PHP 5.6-7.1 much quicker than before.

See https://setup-php.com/i/452

jase-d-ace committed:
pass props to select input

jase-d-ace committed:
move php form to temporary file

jase-d-ace committed:
break up settings into three categories

jase-d-ace committed:
add changes to build folder

jase-d-ace committed:
add caption to each setting

jase-d-ace committed:
add small styling change and change loading message

jase-d-ace committed:
add build files

jase-d-ace committed:
move services file to js directory

jase-d-ace committed:
add build files

jase-d-ace committed:
create css file to eventually use

jblz committed:
Merge in differences in trunk from 2.5.0 release

jblz committed:
Bump version to 2.6.0-alpha

jase-d-ace committed:
small markup changes to components

jase-d-ace committed:
build assets

jase-d-ace committed:
move styles to class-parsely.php for enqueuing

jblz committed:
Build and load admin page styles

jblz committed:
bump package-lock for postcss

jase-d-ace committed:
remove inline styles from markup

jase-d-ace committed:
check in build stuff

jase-d-ace committed:
space out nav tabs

jase-d-ace committed:
more styling work

jase-d-ace committed:
build assets

jase-d-ace committed:
more styling work

jase-d-ace committed:
build assets

jase-d-ace committed:
switch checkbox for toggle switch

jase-d-ace committed:
build assets

jase-d-ace committed:
change toggle color

jase-d-ace committed:
small styling changes

jase-d-ace committed:
build assets

jase-d-ace committed:
add classes for styles

jase-d-ace committed:
styling changes to make label copy bigger

jase-d-ace committed:
build assets

jase-d-ace committed:
fix className spacing

jase-d-ace committed:
build assets

jase-d-ace committed:
fix typo

jase-d-ace committed:
build assets

jase-d-ace committed:
fix widths

jase-d-ace committed:
build assets

jase-d-ace committed:
create SiteDetails component

jase-d-ace committed:
render SiteDetails component

jase-d-ace committed:
build assets

jase-d-ace committed:
changes to SiteDetail and App components

jase-d-ace committed:
styling work

jase-d-ace committed:
build assets

jase-d-ace committed:
create new components for wiping metadata flow

jase-d-ace committed:
render new components in App, and fix SiteDetails styling

jase-d-ace committed:
styling work

jase-d-ace committed:
build assets

jblz commented:
For the checkbox toggles, let's try to use these:
https://github.com/WordPress/gutenberg/blob/19017d79663cff654076612e3c493e95cc5c928d/packages/components/src/toggle-control/README.md

jblz committed:
Add permission check to endpoint
* mv services.js to lib/settings-api-client.js
* apply `current_user_can` via `permission_callback`
* leverage @wordpress/api-fetch
* include api secret in response

jblz committed:
write endpoint and frontend functionality

jblz committed:
Return a 400 error code when error writing settings

jblz committed:
WP_REST_Response all the things

jblz committed:
simplify api client & handle in caller. Call initial fetcher in useEffect.

jblz committed:
actually call update_option in write endpoint

jblz committed:
include package-lock so tests will run

jase-d-ace committed:
add wp-components to enqueue_style

jase-d-ace committed:
install @wordpress/components for FormToggle

jase-d-ace committed:
remove old styles

jase-d-ace committed:
replace old toggle with FormToggle for a11y

jase-d-ace committed:
build assets

jase-d-ace committed:
add font for wipe metadata confirmation flow

jase-d-ace committed:
change confirmation message to apikey

jase-d-ace committed:
build assets

jblz committed:
Merge branch 'remove/jquery-dependency--try-react' of github.com:Parsely/wp-parsely into remove/jquery-dependency--try-react--update-api

jblz committed:
update package-lock

jblz committed:
update build

jblz committed:
Merge pull request #332 from Parsely/remove/jquery-dependency--try-react--update-api

API Improvements for REST-powered Admin page

jblz committed:
prettier & eslint fixes

jase-d-ace committed:
ESLint formatting

jase-d-ace committed:
build assets

jblz commented on src/js/lib/admin-page/components/Setting.js:
```diff
@@ -0,0 +1,47 @@
+		);
+	} else if ( typeof setting[ Object.keys( setting )[ 0 ] ] === 'boolean' ) {
+		input = (
+			<FormToggle
```
Is there a reason you decided to use `FormToggle` vs. `FormControl`? The latter will give us a bit more in the way of accessibility via the [`help` prop](https://github.com/WordPress/gutenberg/blob/e24947929f31b34459a85d31b9c5372f50cdce01/packages/components/src/toggle-control/README.md#help).

jase-d-ace replied:
I was trying both of them, but `FormToggle` accepted a `name` prop that `ToggleControl` didn't, and the lack of `name` stopped my `onChange` function from updating state properly. Is there something I can do to get that to update state?

jase-d-ace replied:
digging into this a little bit deeper, the main reason I can't update state properly when using `ToggleControl` is because `event.target` is undefined when you try to toggle on/off with that. And the reason for that I think is because `ToggleControl` actually renders `FormToggle` under the hood, but doesn't allow the propagation of the event to the child component where `event.target` would actually be defined. Is there a workaround for that?

jblz replied:
Took a stab at reworking all the Settings to use the `Control` variants in #334. Let me know what you think!

jblz committed:
prevent fetchSettings from clobbering compoenent errors

jblz committed:
Declare setting name, pass it back with onChange, leverage WP Control Components

jblz committed:
update build

jase-d-ace committed:
Merge pull request #334 from Parsely/remove/jquery-dependency--try-react--control-components

Declare setting name, pass it back with onChange, leverage WP Control Components

jase-d-ace committed:
add a few styles to modal, and replace lone input with TextControl component

jase-d-ace committed:
styles for modal

jase-d-ace committed:
build assets

jase-d-ace committed:
fix modal styling and button UX

jase-d-ace committed:
fix classes for styling purposes

jase-d-ace committed:
fix and consolidate styles for modal and other small changes

jase-d-ace committed:
build assets

jase-d-ace committed:
add color and style to wipe metadata button

jase-d-ace committed:
fix styling on site details to align with rest of app

jase-d-ace committed:
build assets

jase-d-ace committed:
add styles and eslint fixes

jase-d-ace committed:
build assets

jase-d-ace committed:
add copy to clipboard functionality to SiteDetail component

jase-d-ace committed:
add styles

jase-d-ace committed:
build assets

jase-d-ace committed:
change cursor to pointer for copy CTA

jase-d-ace committed:
build assets

jblz committed:
Format 'pseudo-boolean' settings as strings 'true' or 'false'

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into remove/jquery-dependency--try-react

jblz committed:
update build

jblz committed:
remove explicit postcss dev dependency

jase-d-ace committed:
change UX and styling of modal flow and remove alerts

jase-d-ace committed:
update styles

jase-d-ace committed:
build assets

jase-d-ace committed:
build assets

jase-d-ace committed:
change UX to reflect copied state

jase-d-ace committed:
change default tab to "general"

jase-d-ace committed:
build assets

jase-d-ace committed:
add Notice component and better UX for form submission

jase-d-ace committed:
remove unnecessary packages

jase-d-ace committed:
build assets

jase-d-ace committed:
add failure notice in case of bad form submission

jase-d-ace committed:
build assets

jase-d-ace committed:
eslint fix

jblz committed:
Notices that dismiss themselves, disable form during submit, set values from server, improve error handling

jblz committed:
Merge branch 'remove/jquery-dependency--try-react' of github.com:Parsely/wp-parsely into remove/jquery-dependency--try-react

jblz committed:
update build

jblz commented on src/js/lib/admin-page/style.scss:
```diff
@@ -0,0 +1,166 @@
+	cursor: pointer;
+}
+
+#wp-parsely-react-entrypoint {
```
We should probably wrap most of the above in this qualifier to be sure we're only applying styles to our own elements.

jase-d-ace replied:
Moved!

jase-d-ace replied:
Moved everything into the qualifier! it's all under the `#wp-parsely-react-entry-point` declaration, and everything still looks styled!

jblz commented:
Just noting that I was able to test the error handling by selectively adding lines like this:

```
diff --git a/src/settings-api.php b/src/settings-api.php
index 560aca6..a6850ba 100644
--- a/src/settings-api.php
+++ b/src/settings-api.php
@@ -17,6 +17,7 @@ function wp_parsely_settings_api_permissions_check( WP_REST_Request $request ) {
  * @return WP_REST_Response
  */
 function wp_parsely_settings_api_read_settings( WP_REST_Request $request ) {
+       return new WP_REST_Response( array( 'errors' => [ 'omg' ] ), 400 );
        global $parsely;
        return new WP_REST_Response( $parsely->get_options(), 200 );
 }
@@ -29,6 +30,7 @@ function wp_parsely_settings_api_read_settings( WP_REST_Request $request ) {
  */
 function wp_parsely_settings_api_write_settings( WP_REST_Request $request ) {
        global $parsely, $wp_settings_errors;
+       return new WP_REST_Response( array( 'errors' => [ 'omg' ] ), 400 );

        // `$parsely->validate_options` uses `add_settings_error`
        // TODO: Decouple!
```

jase-d-ace committed:
move all styles underneath #wp-react-entry-point to keep things relegated to only our app

jase-d-ace committed:
style "Undo" button to match "Cancel" button

jase-d-ace committed:
reset form to general after successful submission

jase-d-ace committed:
build assets

jase-d-ace committed:
swap SnackbarList for success Notice to not obscure nav bar on success

jase-d-ace committed:
build assets

jblz commented on package.json:
```diff
@@ -31,7 +31,9 @@
+    "@wordpress/api-fetch": "^4.0.0",
+    "@wordpress/components": "^14.1.1",
     "@wordpress/dom-ready": "^2.13.2",
-    "@wordpress/i18n": "^3.20.0"
```
Looks like we inadvertently removed the i18n dependency.  Can you please add it back and take a pass at wrapping any new user-facing strings for translation?

I'm adding a few suggestions as pointers, but we should ideally wrap them all.

jase-d-ace replied:
I've gone through and wrapped as many strings as I could find in the `__( string, domain )` format. I'm pretty sure I've gotten them all, but let me know if I've missed any

jblz commented on src/js/lib/admin-page/components/App.js:
```diff
@@ -0,0 +1,273 @@
+/**
+ * External dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
```
```suggestion
import { __ } from '@wordpress/i18n';
import { useEffect, useState } from '@wordpress/element';
```

jblz commented on src/js/lib/admin-page/components/App.js:
```diff
@@ -0,0 +1,273 @@
+			( errorFromServer ) => {
+				console.error( errorFromServer );
+				setNotice( {
+					content: 'Could not read settings from the site.',
```
```suggestion
					content: __( 'Could not read settings from the site.', 'wp-parsely' ),
```

jblz commented on src/js/lib/admin-page/components/App.js:
```diff
@@ -0,0 +1,273 @@
+		try {
+			const settingsFromServer = await saveSettingsToServer( settings );
+			setSettings( settingsFromServer );
+			setNotice( { content: 'Settings Successfully Updated!' } );
```
```suggestion
			setNotice( { content: __( 'Settings Successfully Updated!', 'wp-parsely' ) } );
```

jblz commented on src/js/lib/admin-page/components/App.js:
```diff
@@ -0,0 +1,273 @@
+	const nav = (
+		<nav className="controls">
+			<div className="nav-control" onClick={ () => _setCurrentTab( 'general' ) }>
+				<span className={ `${ currentTab === 'general' ? 'active' : '' }` }>General</span>
```
```suggestion
				<span className={ `${ currentTab === 'general' ? 'active' : '' }` }>{ __( 'General', 'wp-parsely' ) }</span>
```

jblz commented on src/js/lib/admin-page/index.js:
```diff
@@ -1,7 +1,13 @@
 import domReady from '@wordpress/dom-ready';
 import { __, sprintf } from '@wordpress/i18n';
```
Now that we're completely taking over the admin panel, we can pull out this line:
```suggestion
```

...& also the remainder of this function:
https://github.com/Parsely/wp-parsely/blob/171714f56d90a57f8385dfd74965d775fc5846cc/src/js/lib/admin-page/index.js#L10-L33

jase-d-ace commented:
I can't find the button to mark this "Ready for Review" anymore. @jblz who should I ask to look through?

jblz commented on src/js/lib/admin-page/index.js:
```diff
@@ -1,7 +1,13 @@
 import domReady from '@wordpress/dom-ready';
 import { __, sprintf } from '@wordpress/i18n';
 
+import App from './components/App';
```
Let's make the entry component name and path more descriptive than `App` so it's easier to grok in the editor. Maybe something like:

```suggestion
import ParselySettings from './components/ParselySettings';
```

(we'll need to `git mv ./components/App.js ./components/ParselySettings.js`, of course)

jase-d-ace replied:
This is done in the latest couple of commits

jblz commented:
> I can't find the button to mark this "Ready for Review" anymore. @jblz who should I ask to look through?

You'll want to really flesh out the PR description (you can pull in the [PR template](https://github.com/Parsely/wp-parsely/blob/5383a314eea04979171d02d1c7cf8792977e1b34/.github/pull_request_template.md) and go from there).  Make sure to write a comprehensive test plan so someone who isn't that familiar with the dev env and plugin in general could provide meaningful feedback.

...then I'd add two labels:
* `[Status] Needs Review`
* `[Status] Needs Design Review`

jblz commented on src/class-parsely.php:
```diff
@@ -1623,7 +1625,7 @@ private function get_categories( $post_id, $delimiter = '/' ) {
 	 * on variables.
 	 */
-	private function get_options() {
+	public function get_options() {
```
@hbbtstar @GaryJones -- any qualms about making this public? Can you think of a better way to get this data in the endpoint here?

https://github.com/Parsely/wp-parsely/blob/171714f56d90a57f8385dfd74965d775fc5846cc/src/settings-api.php#L21

GaryJones replied:
Not at this point, no. In a future iteration where the Parsely class didn't exist, then we might be more careful about how we share and access settings, but for this PR, go ahead and make it public.

jblz commented on src/settings-api.php:
```diff
@@ -0,0 +1,70 @@
+
+	update_option( Parsely::OPTIONS_KEY, $validated_settings );
+
+	return new WP_REST_Response( $validated_settings, 200 );
```
We should apply the same filters on this returned data as the `GET` endpoint:

```suggestion
	return new WP_REST_Response( $parsely->get_options(), 200 );
```

jblz commented on src/settings-api.php:
```diff
@@ -0,0 +1,70 @@
+		return new WP_REST_Response( array( 'errors' => $wp_settings_errors ), 400 );
+	}
+
+	update_option( Parsely::OPTIONS_KEY, $validated_settings );
```
We should return an error when `update_option` fails, e.g.:

```suggestion
	$result = update_option( Parsely::OPTIONS_KEY, $validated_settings );
	if ( ! $result ) {
		return new WP_REST_Response( array( 'errors' => __( 'Unable to update settings', 'wp-parsely' ) ), 400 );
	}
```

jase-d-ace committed:
rename file and wrap strings in i18n function

jase-d-ace committed:
change mounting component name from App to ParselySettings

jase-d-ace committed:
add @wordpress/i18n package

jase-d-ace committed:
build assets

jase-d-ace committed:
remove unnecessary file

jase-d-ace committed:
add i18n support for all customer-facing strings

jase-d-ace committed:
build assets

jase-d-ace commented:
> > I can't find the button to mark this "Ready for Review" anymore. @jblz who should I ask to look through?
> 
> You'll want to really flesh out the PR description (you can pull in the [PR template](https://github.com/Parsely/wp-parsely/blob/5383a314eea04979171d02d1c7cf8792977e1b34/.github/pull_request_template.md) and go from there). Make sure to write a comprehensive test plan so someone who isn't that familiar with the dev env and plugin in general could provide meaningful feedback.
> 
> ...then I'd add two labels:
> 
> * `[Status] Needs Review`
> * `[Status] Needs Design Review`

It looks like you're the author of this PR. Should I just go ahead and edit your top comment with all of the PR template stuff? Or should we make me the new author of this?

jblz commented:
> It looks like you're the author of this PR. Should I just go ahead and edit your top comment with all of the PR template stuff? Or should we make me the new author of this?

I don't think Github allows changing the PR author (as ridiculous as that sounds).  Feel free to either edit the description directly or create a new branch / PR with these contents.

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+						value={ settings.track_post_types }
+						label={ __( 'Track Post Types', 'wp-parsely' ) }
+						onChange={ handleInputChange }
+						note={ __( '....', 'wp-parsely' ) }
```
Is this translatable string still to be entered?

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+						value={ settings.track_page_types }
+						label="Track Page Types"
+						onChange={ handleInputChange }
+						note="...."
```
Is this untranslatable string still to be entered?

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+					<Setting
+						name="track_page_types"
+						value={ settings.track_page_types }
+						label="Track Page Types"
```
Untranslated string here.

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+					<Setting
+						name="disable_javascript"
+						value={ settings.disable_javascript }
+						label={ __( 'Disable Javascript', 'wp-parsely' ) }
```
```suggestion
						label={ __( 'Disable JavaScript', 'wp-parsely' ) }
```

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+						value={ settings.disable_javascript }
+						label={ __( 'Disable Javascript', 'wp-parsely' ) }
+						onChange={ handleInputChange }
+						note={ __( 'Default: Off. Disable our javascript tracking if you use a separate system for JS tracking', 'wp-parsely' ) }
```
```suggestion
						note={ __( 'Default: Off. Disable the Parse.ly plugin JavaScript tracking if you use a separate system for JavaScript tracking', 'wp-parsely' ) }
```

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+						value={ settings.disable_amp }
+						label={ __( 'Disable AMP', 'wp-parsely' ) }
+						onChange={ handleInputChange }
+						note={ __( 'Default: On. Disable our AMP tracking if you use a separate system to track AMP content', 'wp-parsely' ) }
```
```suggestion
						note={ __( 'Default: On. Disable Parse.ly plugin AMP tracking if you use a separate system to track AMP content', 'wp-parsely' ) }
```

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+						apikey={ settings.apikey }
+						postsToTrack={ settings.track_post_types }
+						pagesToTrack={ settings.track_page_types }
+						pluginVersion={ __('2.5', 'wp-parsely' ) }
```
Can this version number be grabbed from somewhere else, rather than having yet another place to remember to update it?

Also, it doesn't need translating.

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+						postsToTrack={ settings.track_post_types }
+						pagesToTrack={ settings.track_page_types }
+						pluginVersion={ __('2.5', 'wp-parsely' ) }
+						phpVersion={ __( '7.4.1', 'wp-parsely' ) }
```
PHP version doesn't need translating.

GaryJones commented on src/js/lib/admin-page/components/ParselySettings.js:
```diff
@@ -0,0 +1,272 @@
+				disabled={ updatingSettings }
+				type="submit"
+				className="button-primary"
+				value={ __( 'do the thing!', 'wp-parsely' ) }
```
Can we come up with a more sensible value here? Check other WP default admin screens to see what their submit button values are.

GaryJones commented on src/js/lib/admin-page/components/Setting.js:
```diff
@@ -0,0 +1,52 @@
+		input = <ToggleControl name={ name } label={ label } onChange={ _onChange } checked={ value } />;
+	} else if ( [ 'track_post_types', 'track_page_types' ].includes( name ) ) {
+		input = (
+			<SelectControl
```
Multi-select elements are notoriously inaccessible - can we find a way to drop them completely?

GaryJones commented on src/js/lib/admin-page/components/SiteDetails.js:
```diff
@@ -0,0 +1,50 @@
+				</div>
+				<div className="details-info">
+					<span className="details-info-item">
+						{ __( 'Parsely Site ID: ', 'wp-parsely' ) } { apikey }
```
Don't include the trailing space in the translatable string, here, or any of the instances below.

GaryJones commented on src/js/lib/admin-page/components/WipeMetadataCache.js:
```diff
@@ -0,0 +1,20 @@
+	return (
+		<div className="container">
+			<div className="wipe-metadata-container">
+				<label className="wipe-metadata-label">{ __( label, 'wp-parsely' ) }</label>
```
I know for PHP, you can't call a translation function with the first arg as a variable - it should be a literal string instead, made translatable at the point that string is defined.

I don't know how the JS equivalent of `__()` works, but it may be the same, and if so, this instance doesn't need the translation function.

GaryJones commented on src/js/lib/admin-page/components/Setting.js:
```diff
@@ -0,0 +1,52 @@
+			<div className="setting-item">
+				<div className="setting-item--control">
+					{ input }
+					<p className="subtext">{ __( note, 'wp-parsely' ) }</p>
```
See other note about passing a variable to the translation function. I bet the string that should appear in `note` variable is already translated elsewhere?

GaryJones commented on src/js/lib/admin-page/components/WipeMetadataCache.js:
```diff
@@ -0,0 +1,20 @@
+				<label className="wipe-metadata-label">{ __( label, 'wp-parsely' ) }</label>
+				<div className="wipe-metadata-controls">
+					<button type="button" className="wipe-metadata-button" onClick={ () => onClick( true ) }>{ __( 'Wipe your metadata?', 'wp-parsely' ) }</button>
+					{ __( note, 'wp-parsely' ) }
```
Another translated variable that shouldn't be.

GaryJones commented on src/js/lib/admin-page/components/WipeMetadataCache.js:
```diff
@@ -0,0 +1,20 @@
+			<div className="wipe-metadata-container">
+				<label className="wipe-metadata-label">{ __( label, 'wp-parsely' ) }</label>
+				<div className="wipe-metadata-controls">
+					<button type="button" className="wipe-metadata-button" onClick={ () => onClick( true ) }>{ __( 'Wipe your metadata?', 'wp-parsely' ) }</button>
```
```suggestion
					<button type="button" className="wipe-metadata-button" onClick={ () => onClick( true ) }>{ __( 'Wipe Parse.ly plugin metadata?', 'wp-parsely' ) }</button>
```

Also, what benefit does the onClick attribute value provide here?

GaryJones commented on src/js/lib/admin-page/components/WipeMetadataCache.js:
```diff
@@ -0,0 +1,20 @@
+					<button type="button" className="wipe-metadata-button" onClick={ () => onClick( true ) }>{ __( 'Wipe your metadata?', 'wp-parsely' ) }</button>
+					{ __( note, 'wp-parsely' ) }
+					<div className="wipe-metadata-warning">
+						{ setting.parsely_wipe_metadata_cache ? <h1>{ __( 'DANGER: METADATA WILL BE WIPED UPON FORM SUBMISSION', 'wp-parsely' ) }</h1> : '' }
```
Why is `h1` being used? It's not the title of the admin page. If it's an important paragraph, then visually style it as so, without using incorrect semantics.

GaryJones commented on src/js/lib/admin-page/components/WipeMetadataModal.js:
```diff
@@ -0,0 +1,40 @@
+			<div onClick={ () => setFlag( false ) } className="modal-background">
+			</div>
+			<div className="modal">
+				<strong>{ __( 'Type', 'wp-parsely' ) } <span className="code">{ apikey }</span> { __( 'below and then click Confirm if you really want to delete all stored metadata. This action cannot be undone', 'wp-parsely' ) }</strong>
```
Does the JS `__()` support placeholders? Right now, this is split into separate strings, which makes it harder for translators:

- `'Type'`
- `'below and then [...]'`

With a placeholder, then it can be a single string:

- `'Type %s below and then [...]'`
...with the `%s` then replaced with `<code>{ apikey }</code>` (not sure why a span with a class name is used instead of a `code` element) after the translation has potentially happened.
...and add a `/* translators: API Key */` translator's comment to tell the translator what the placeholder will contain.

GaryJones commented on src/js/lib/admin-page/style.scss:
```diff
@@ -0,0 +1,165 @@
+
+#wp-parsely-react-entrypoint {
+	.components-snackbar {
+		border-left: 5px solid green;
+		&.settings-error {
+			border-left: 5px solid red;
```
Instead of just `green` and `red`, can you find the proper hex codes that match up to the wp-admin colour schemes? Let's try and make it fit in as much as possible.

GaryJones commented on src/js/lib/admin-page/style.scss:
```diff
@@ -0,0 +1,165 @@
+		}
+	}
+	#wp-parsely_version {
+		color: $gray;
```
Will this grey text work on the various different wp-admin colour schemes and give sufficient colour contrast?

GaryJones commented on src/js/lib/admin-page/style.scss:
```diff
@@ -0,0 +1,165 @@
+		padding-bottom: 1em;
+	}
+
+	.code {
+		font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New', monospace;
+	}
```
Switch to `<code>` element, and let the browser or wp-admin default handle the styling of this.

GaryJones commented on src/settings-api.php:
```diff
@@ -0,0 +1,70 @@
+ * @param WP_REST_Request $request Full data about the request.
+ * @return bool
+ */
+function wp_parsely_settings_api_permissions_check( WP_REST_Request $request ) {
```
No need to include this arg if the logic isn't using it.

GaryJones commented on src/settings-api.php:
```diff
@@ -0,0 +1,70 @@
+ * @param WP_REST_Request $request Full data about the request.
+ * @return WP_REST_Response
+ */
+function wp_parsely_settings_api_read_settings( WP_REST_Request $request ) {
```
Same here.

GaryJones commented on src/settings-api.php:
```diff
@@ -0,0 +1,70 @@
+	return new WP_REST_Response( $validated_settings, 200 );
+}
+
+function wp_parsely_settings_api_init() {
```
Needs some docs.

jase-d-ace committed:
missed a couple of i18n strings, and other small changes

jase-d-ace committed:
additional styles for copy CTA

jase-d-ace committed:
styling changes, semantic tag changes, and other fixes to wipe metadata flow

jase-d-ace committed:
build assets

pauarge commented:
Although we'll be working on the settings page, we don't want to add React. It is a very heavy dependency and we can achieve what we want without it. 

This PR has become stale and has many conflicts, so will close it but it still remains a good reference point to check when implementing the settings page redesign.
